torch/csrc/jit/tensorexpr/ir_serializer.cpp

```diff
@@ -0,0 +1,557 @@
+      v,
+      {
+          {"base_handle", visitAndSerialize(v->base_handle())},
+          {"indices", visitAndSerialize(v->indices())},
+          {"mask", visitAndSerialize(v->mask())},
+          {"stmt_type", "Load"},
+      });
+}
+
+void IRSerializer::visit(const Store* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"stmt_type", "Store"},
+          {"indices", visitAndSerialize(v->indices())},
+          {"value", visitAndSerialize(v->value())},
+          {"mask", visitAndSerialize(v->mask())},
+          {"buf", visitAndSerialize(v->buf())},
+      });
+}
+
+void IRSerializer::visit(const Block* v) {
+  JSON_CACHE_GUARD();
+  // TODO: why is Block using std::list instead of std::vector ?
+  std::vector<json> stmts;
+  for (Stmt* s : *v) {
+    s->accept(this);
+    stmts.push_back(JSONOf(s));
+  }
+  putJSON(
+      v,
+      {
+          {"stmt_type", "Block"},
+          {"stmts", stmts},
+      });
+}
+
+void IRSerializer::visit(const For* v) {
+  JSON_CACHE_GUARD();
+  auto buffer_mapping = v->loop_options().get_buffer_mapping();
+  json buffer_mapping_json;
+  for (const auto& mapping : buffer_mapping) {
+    buffer_mapping_json[mapping.first] = visitAndSerialize(mapping.second);
+  }
+  json loop_options_json = {
+      {"buffer_mapping", buffer_mapping_json},
+      {"gpu_block_index", v->loop_options().gpu_block_index()},
+      {"gpu_thread_index", v->loop_options().gpu_thread_index()},
+  };
+  json for_json = {
+      {"stmt_type", "For"},
+      {"var", visitAndSerialize(v->var())},
+      {"start", visitAndSerialize(v->start())},
+      {"stop", visitAndSerialize(v->stop())},
+      {"loop_options", loop_options_json},
+  };
+  if (v->body()) {
+    for_json["body"] = visitAndSerialize(v->body());
+  }
+  putJSON(v, for_json);
+}
+
+void IRSerializer::visit(const Broadcast* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "Broadcast"},
+          {"value", visitAndSerialize(v->value())},
+          {"lanes", v->lanes()},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void IRSerializer::visit(const IfThenElse* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "IfThenElse"},
+          {"true_value", visitAndSerialize(v->true_value())},
+          {"false_value", visitAndSerialize(v->false_value())},
+          {"condition", visitAndSerialize(v->condition())},
+          {"dtype", serializeDtype(v->dtype())},
+      });
+}
+
+void IRSerializer::visit(const BaseCallNode* v) {
+  // should have visited the inherited class
+  assert(false);
+}
+
+void IRSerializer::visit(const Intrinsics* v) {
+  putJSON(
+      v,
+      {{"expr_type", "Intrinsics"},
+       {"dtype", serializeDtype(v->dtype())},
+       {"func_name", v->func_name()},
+       {"params", visitAndSerialize(v->params())}});
+}
+
+void IRSerializer::visit(const Allocate* v) {
+  JSON_CACHE_GUARD();
+  // TODO: why is allocate a stmt and not an expr ?
```

Comment:
Think of it as `int A[size];` not `malloc(size)`.

We could make it and Expr, but I'd want to restrict it to appearing only on the RHS of a `Let` statement, since I don't know of any reason we'd take the added complexity of handing allocation in the RHS of Stores, or worse internal For stopping conditions and the like. In that case I don't see the advantage of it being an Expr.

Reply:
Thanks, makes sense. yea most of these comments were just for my own understanding...

Reply:
No worries, happy to answer questions in whichever format :)